pkg/scheduler/schedule_one.go

```diff
@@ -336,6 +263,46 @@ func (sched *Scheduler) bindingCycle(ctx context.Context, state *framework.Cycle
 		// Unlike the logic in schedulingCycle(), we don't bother deleting the entries
 		// as `podsToActivate.Map` is no longer consumed.
 	}
+
+	return nil
+}
+
+func (sched *Scheduler) handleBindingCycleError(
+	ctx context.Context,
+	state *framework.CycleState,
+	fwk framework.Framework,
+	podInfo *framework.QueuedPodInfo,
+	start time.Time,
+	scheduleResult ScheduleResult,
+	status *framework.Status) {
+
+	assumedPod := podInfo.Pod
+	// trigger un-reserve plugins to clean up state associated with the reserved Pod
+	fwk.RunReservePluginsUnreserve(ctx, state, assumedPod, scheduleResult.SuggestedHost)
+	if forgetErr := sched.Cache.ForgetPod(assumedPod); forgetErr != nil {
+		klog.ErrorS(forgetErr, "scheduler cache ForgetPod failed")
+	} else {
+		// "Forget"ing an assumed Pod in binding cycle should be treated as a PodDelete event,
+		// as the assumed Pod had occupied a certain amount of resources in scheduler cache.
+		//
+		// Avoid moving the assumed Pod itself as it's always Unschedulable.
+		// It's intentional to "defer" this operation; otherwise MoveAllToActiveOrBackoffQueue() would
+		// update `q.moveRequest` and thus move the assumed pod to backoffQ anyways.
+		if status.IsUnschedulable() {
+			defer sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, func(pod *v1.Pod) bool {
+				return assumedPod.UID != pod.UID
+			})
+		} else {
+			sched.SchedulingQueue.MoveAllToActiveOrBackoffQueue(internalqueue.AssignedPodDelete, nil)
+		}
+	}
+
+	reason := v1.PodReasonSchedulerError
+	if status.IsUnschedulable() {
+		reason = v1.PodReasonUnschedulable
+	}
+
+	sched.FailureHandler(ctx, fwk, podInfo, status.AsError(), reason, clearNominatedNode, start)
```

Comment:
Food for thought:

kind of weird that we have to pass an error and a reason.

Perhaps we should be passing a status (although I know it's a bit more tricky for scheduling cycle), or the error should contain information about whether it's unschedulable or not (by implementing `Is`)

But this can be left for follow up, if you have extra time.

Reply:
It's something we can optimize later. I'll follow this when I have time.